posts/2018-11-28-http4s-error-handling-mtl-2.md

```diff
@@ -0,0 +1,459 @@
+---
+layout: post
+title: Error handling in Http4s with classy optics – Part 2
+
+meta:
+  nav: blog
+  author: gvolpe
+  pygments: true
+
+tut:
+  scala: 2.12.7
+  binaryScala: "2.12"
+  scalacOptions:
+    - -Ypartial-unification
+  plugins:
+    - org.spire-math::kind-projector:0.9.9
+  dependencies:
+    - org.scala-lang:scala-library:2.12.7
+    - org.typelevel::cats-core:1.4.0
+    - org.typelevel::cats-effect:1.1.0-M1
+    - org.http4s::http4s-blaze-server:0.20.0-M3
+    - org.http4s::http4s-circe:0.20.0-M3
+    - org.http4s::http4s-dsl:0.20.0-M3
+    - io.circe::circe-core:0.10.0
+    - io.circe::circe-generic:0.10.0
+    - co.fs2::fs2-core:1.0.0
+    - com.olegpy::meow-mtl:0.2.0
+    - com.chuusai::shapeless:2.3.3
+
+---
+
+This is a continuation of my [previous blog post](https://typelevel.org/blog/2018/08/25/http4s-error-handling-mtl.html). Make sure you have read that one before continuing here.
+
+I recently gave a 20 minutes talk on `classy optics` at the unconference of [Scale by the Bay](http://scale.bythebay.io/) where I also talked about this error handling technique and on my way back home I was still thinking of different ways of doing this. So, after some exploratory work, I came up with a few different alternatives.
+
+### Issues with first approach
+
+Something that made me cringe and that a few of my colleagues at work were not happy with was that the algebras had no association with the error type defined in `HttpErrorHandler[F, E]` so the type-safety was down to the programmer's discipline and in this case the compiler was not able to do much.
+
+When working with `EitherT[F, E, A]` or a bifunctor `IO[E, A]` we have a clear error type whereas by just relying on a single `F[A]` with a `MonadError[F, Throwable]` instance we lose this property. However, most of the time we only care about mapping a few business errors to Http responses if working on a REST API.
+
+#### Data types with error handling capabilities
+
+As of now I still prefer the latter for a number of reasons. Here's a comparison table of all the different approaches I can think of:
+
+|            | EitherT[IO, E, ?] | IO[E, A] | IO[A] | IO[A] + Classy Optics |
+|------------|-------------------|----------|-------|-----------------------|
+| Concise code style | ✘ | ✔ | ✔ | ✔ |
+| Polymorphic code | ✔ | ✘ | ✔ | ✔ |
+| Performance overhead | **2x** | ✔ | ✔ | ✔ |
```

Comment:
What happens in the `IO` almost always dominates the overhead of `EitherT`.  The benchmarks that make monad transformers look bad tend to be a worst case.

Reply:
Would `~2x` be better? It's more an estimation than a concrete result given the extra depth of `flatMap` calls involved when using Monad Transformers.